File: frontend/lib/search-disclosure.ts
import { WORK_FIELDS, validWorkFields, validQueryMode, type QueryMode } from "./search-evidence.ts";
import type { SearchRanking } from "./search-ranking.ts";
import { validContentKinds } from "./transcript-segments.ts";
import type { TermDiagnostic } from "./search-diagnostics.ts";
import { SEARCH_DATE_FIELDS, dateBoundsMatch, validDateBounds, validDiagnosticsMode, type DiagnosticsMode } from "./search-exploration.ts";
import { exactKeys, objectValue, sameUuid, validUuid } from "./wire-guards.ts";

export const SEARCH_DISCLOSURE_FIELDS = ["diagnostics", "applied_filters", "query_interpretation", "warnings"] as const;
const sources = ["work_items", "artifacts", "transcripts"] as const;
type Source = typeof sources[number];
export const DEFAULT_SEARCH_FILTERS = {
  work_items: { work_fields: [...WORK_FIELDS], status: "all", tag: null, source_client: null, source_session_id: null, external_url: null, duplicate_scope: "canonical", canonical_work_item_id: null, view: "full" },
  artifacts: { artifact_id: null, work_item_id: null, include_deleted: false, sensitive: null, mime_type: null, created_by_agent_session_id: null },
  transcripts: { content_kinds: null, work_item_id: null, agent_session_id: null, client: null, kind: null, status: null }
};
export type AppliedSearchFilters = { project_id: string; project_ids: null } & Record<Source, Record<string, unknown> | null>;
export type QueryInterpretation = { q: string; query_mode: QueryMode } & Record<Source, { match_mode: string; fields: string[]; fulltext: boolean | null } | null>;
export type SearchDisclosure = {
  diagnostics: DiagnosticsMode;
  applied_filters: AppliedSearchFilters;
  query_interpretation: QueryInterpretation;
  warnings: never[];
};

function validFilter(source: Source, value: unknown): boolean {
  const row = objectValue(value);
  if (!row || !exactKeys(row, [...Object.keys(DEFAULT_SEARCH_FILTERS[source]), ...(source === "artifacts" && "semantic" in row ? ["semantic"] : []), ...SEARCH_DATE_FIELDS.filter((key) => key in row)]) || !validDateBounds(row, true)) return false;
  return Object.entries(row).every(([key, value]) => {
    if (SEARCH_DATE_FIELDS.includes(key as typeof SEARCH_DATE_FIELDS[number])) return value !== null;
    if (key === "semantic") return value === true;
    if (value === null) return !["status", "view", "duplicate_scope", "include_deleted"].includes(key) || source === "transcripts" && key === "status";
    if (key === "work_fields") return validWorkFields(value);
    if (key === "content_kinds") return validContentKinds(value);
    if (["work_item_id", "artifact_id", "canonical_work_item_id"].includes(key)) return validUuid(value);
    if (["include_deleted", "sensitive"].includes(key)) return typeof value === "boolean";
    if (key === "status") return (source === "work_items" ? ["all", "pending", "active", "to-review", "dropped", "deferred", "done", "wont-do", "promoted"] : ["waiting", "pending", "processing", "ready", "failed"]).includes(String(value));
    if (key === "view") return value === "full" || value === "roots";
    if (key === "duplicate_scope") return ["canonical", "aliases", "all"].includes(String(value));
    if (key === "kind") return ["primary", "subagent", "imported"].includes(String(value));
    const maximum = key === "tag" ? 50 : ["client", "source_client"].includes(key) ? 80 : key === "external_url" ? 2000 : key === "mime_type" ? 255 : 200;
    return typeof value === "string" && Array.from(value).length <= maximum;
  });
}

export function decodeSearchDisclosure(value: unknown, projectId: string, searched: Source[]): SearchDisclosure {
  const page = objectValue(value);
  const applied = objectValue(page?.applied_filters);
  const interpreted = objectValue(page?.query_interpretation);
  if (!validDiagnosticsMode(page?.diagnostics) || !applied || !exactKeys(applied, ["project_id", "project_ids", ...sources]) || !sameUuid(applied.project_id, projectId) || applied.project_ids !== null
    || !interpreted || !exactKeys(interpreted, ["q", "query_mode", ...sources]) || !validQueryMode(interpreted.query_mode) || typeof interpreted.q !== "string" || Array.from(interpreted.q).length > 1000 || interpreted.q !== interpreted.q.trim()
    || !Array.isArray(page?.warnings)) throw new Error("Mnemonic returned invalid search disclosure.");
  for (const source of sources) {
    if (!searched.includes(source)) {
      if (applied[source] !== null || interpreted[source] !== null) throw new Error("Mnemonic disclosed an unsearched source as searched.");
      continue;
    }
    const interpretation = objectValue(interpreted[source]);
    if (!validFilter(source, applied[source]) || !interpretation || !exactKeys(interpretation, ["match_mode", "fields", "fulltext"])) throw new Error("Mnemonic returned invalid search interpretation.");
    const isWork = source === "work_items";
    const semanticArtifact = source === "artifacts" && objectValue(applied[source])?.semantic === true;
    const fields = semanticArtifact ? ["content"] : isWork ? objectValue(applied[source])?.work_fields : interpretation.fulltext ? ["metadata", "content"] : ["metadata"];
    const constrained = interpreted.query_mode !== "terms" || interpreted.q.includes('"');
    const modes = semanticArtifact ? ["semantic_passages"] : !interpreted.q ? ["browse"] : interpreted.query_mode === "literal" ? ["literal"] : constrained ? [isWork ? "postgresql_phrase_terms" : "phrase"] : isWork ? ["postgresql_plain_terms_or_substring", "hybrid_lexical_semantic"] : ["all_terms"];
    if (semanticArtifact && (!interpreted.q || constrained || interpretation.fulltext !== true)) throw new Error("Mnemonic returned invalid artifact semantic interpretation.");
    if (!modes.includes(String(interpretation.match_mode)) || (isWork ? interpretation.fulltext !== null : typeof interpretation.fulltext !== "boolean")
      || !Array.isArray(interpretation.fields) || JSON.stringify(interpretation.fields) !== JSON.stringify(fields)) throw new Error("Mnemonic returned invalid search interpretation.");
  }
  if (page.warnings.length !== 0) throw new Error("Mnemonic returned invalid search warnings.");
  return { diagnostics: page.diagnostics as DiagnosticsMode, applied_filters: applied as AppliedSearchFilters, query_interpretation: interpreted as QueryInterpretation, warnings: page.warnings as SearchDisclosure["warnings"] };
}

export function validateSearchDisclosure(disclosure: SearchDisclosure, source: Source, expected: Record<string, unknown>, fulltext?: boolean, query?: string, queryMode: QueryMode = "terms"): void {
  const applied = disclosure.applied_filters[source];
  const interpretation = disclosure.query_interpretation[source];
  if (disclosure.query_interpretation.query_mode !== queryMode || query !== undefined && disclosure.query_interpretation.q !== query.trim()
    || applied && !dateBoundsMatch(applied, expected)
    || applied && Object.entries(expected).some(([key, value]) => ["work_item_id", "artifact_id", "canonical_work_item_id"].includes(key) && value != null ? !sameUuid(applied[key], value) : SEARCH_DATE_FIELDS.includes(key as typeof SEARCH_DATE_FIELDS[number]) ? false : ["content_kinds", "work_fields"].includes(key) ? JSON.stringify(applied[key]) !== JSON.stringify(value) : applied[key] !== value)
    || interpretation && fulltext !== undefined && interpretation.fulltext !== fulltext) throw new Error("Mnemonic returned search disclosure outside the requested scope.");
}

export type FullWorkSearchDetail = SearchRanking & { detail: "full"; work_rank_scope: "work_items"; term_diagnostics: TermDiagnostic[] };

File: backend/src/mnemonic_api/search_disclosure.py
"""Bounded, source-specific descriptions of the search that actually ran."""

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from mnemonic_api.search_diagnostics import SearchFacet
from mnemonic_api.search_exploration_schemas import DateBounds, DiagnosticsMode
from mnemonic_api.search_projects import ProjectSelection, selected_project_ids
from mnemonic_api.search_query import QueryMode, parse_query
from mnemonic_api.transcript_normalization import ContentKind
from mnemonic_api.work_search_fields import WORK_FIELDS, WorkFields


class DisclosureModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class WorkAppliedFilters(DisclosureModel, DateBounds):
    status_scope: Literal["effective", "work_item"] = Field(
        default="effective", exclude_if=lambda value: value == "effective",
    )
    work_fields: WorkFields = Field(default_factory=lambda: list(WORK_FIELDS))
    status: Literal[
        "pending", "active", "to-review", "dropped", "deferred", "done",
        "wont-do", "promoted", "all",
    ] = "all"
    tag: str | None = None
    source_client: str | None = None
    source_session_id: str | None = None
    external_url: str | None = None
    duplicate_scope: Literal["canonical", "aliases", "all"] = "canonical"
    canonical_work_item_id: UUID | None = None
    view: Literal["full", "roots"] = "full"


class ArtifactAppliedFilters(DisclosureModel, DateBounds):
    semantic: bool = Field(default=False, exclude_if=lambda value: value is False)
    artifact_id: UUID | None = None
    work_item_id: UUID | None = None
    include_deleted: bool = False
    sensitive: bool | None = None
    mime_type: str | None = None
    created_by_agent_session_id: str | None = None


class TranscriptAppliedFilters(DisclosureModel, DateBounds):
    content_kinds: list[ContentKind] | None = Field(default=None, min_length=1, max_length=8)
    work_item_id: UUID | None = None
    agent_session_id: str | None = None
    client: str | None = None
    kind: Literal["primary", "subagent", "imported"] | None = None
    status: Literal["waiting", "pending", "processing", "ready", "failed"] | None = None


class AppliedSearchFilters(DisclosureModel):
    project_id: UUID | None
    project_ids: list[UUID] | None = Field(default=None, min_length=1, max_length=10)
    # Null identifies an unsearched source, including a disabled artifact library.
    work_items: WorkAppliedFilters | None = None
    artifacts: ArtifactAppliedFilters | None = None
    transcripts: TranscriptAppliedFilters | None = None


class SourceQueryInterpretation(DisclosureModel):
    match_mode: Literal[
        "browse", "postgresql_plain_terms_or_substring", "hybrid_lexical_semantic", "all_terms",
        "postgresql_phrase_terms", "phrase", "literal",
        "semantic_passages",
    ]
    fields: list[Literal[
        "title", "summary", "tags", "checkpoint", "identifiers", "provenance",
        "metadata", "content",
    ]] = Field(max_length=6)
    # Work lexical fields, including checkpoint prose, are controlled by work_fields.
    fulltext: bool | None


class QueryInterpretation(DisclosureModel):
    query_mode: QueryMode = "terms"
    q: str = Field(max_length=1000)
    work_items: SourceQueryInterpretation | None = None
    artifacts: SourceQueryInterpretation | None = None
    transcripts: SourceQueryInterpretation | None = None


class SearchWarning(DisclosureModel):
    code: Literal["phrase_operators_ignored", "sensitive_content_withheld"] = (
        "phrase_operators_ignored"
    )
    sources: list[SearchFacet] = Field(min_length=1, max_length=3)
    message: Literal[
        "Quoted phrases are not supported; quotation marks do not require adjacent words.",
        "Sensitive artifact contents were withheld; zero matches do not establish absence."
    ] = "Quoted phrases are not supported; quotation marks do not require adjacent words."


class SearchDisclosure(DisclosureModel):
    diagnostics: DiagnosticsMode = "on_empty"
    applied_filters: AppliedSearchFilters
    query_interpretation: QueryInterpretation
    warnings: list[SearchWarning] = Field(max_length=2)

    @model_validator(mode="after")
    def disclose_withholding(self) -> SearchDisclosure:
        self.warnings = [warning for warning in self.warnings
                         if warning.code != "sensitive_content_withheld"]
        self.warnings += withholding_warnings(self)
        return self


def withholding_warnings(page) -> list[SearchWarning]:
    coverage = getattr(page, "coverage", None)
    count = getattr(page, "sensitive_content_withheld", 0)
    if coverage is not None:
        count = coverage.artifacts.sensitive_content_withheld
    return [SearchWarning(
        code="sensitive_content_withheld", sources=["artifacts"],
        message="Sensitive artifact contents were withheld; zero matches do not establish absence.",
    )] if count else []


def search_disclosure(
    project_id: ProjectSelection, q: str | None, *,
    work_items: WorkAppliedFilters | None = None,
    artifacts: ArtifactAppliedFilters | None = None,
    transcripts: TranscriptAppliedFilters | None = None,
    semantic: bool = False, fulltext: bool = False, query_mode: QueryMode = "terms",
    diagnostics: DiagnosticsMode = "on_empty",
) -> SearchDisclosure:
    query = (q or "").strip()
    applied = AppliedSearchFilters(
        project_id=project_id if isinstance(project_id, UUID) else None,
        project_ids=(None if isinstance(project_id, UUID)
                     else list(selected_project_ids(project_id))),
        work_items=work_items, artifacts=artifacts, transcripts=transcripts,
    )
    intent = parse_query(query, query_mode)
    interpretation = QueryInterpretation(q=query, query_mode=query_mode)
    sources: list[SearchFacet] = []
    if work_items is not None:
        sources.append("work_items")
        mode = "hybrid_lexical_semantic" if semantic else "postgresql_plain_terms_or_substring"
        if intent.constrained:
            mode = "literal" if query_mode == "literal" else "postgresql_phrase_terms"
        interpretation.work_items = SourceQueryInterpretation(
            match_mode=mode if query else "browse", fulltext=None,
            fields=list(work_items.work_fields),
        )
    for facet, filters in (("artifacts", artifacts), ("transcripts", transcripts)):
        if filters is None:
            continue
        sources.append(facet)
        content_mode = ("literal" if query_mode == "literal" else
                        "phrase" if intent.constrained else "all_terms")
        source = SourceQueryInterpretation(
            match_mode=content_mode if query else "browse", fulltext=fulltext,
            fields=["metadata", "content"] if fulltext else ["metadata"],
        )
        if (facet == "artifacts" and isinstance(filters, ArtifactAppliedFilters)
                and filters.semantic):
            source.match_mode = "semantic_passages"
            source.fields = ["content"]
        setattr(interpretation, facet, source)
    return SearchDisclosure(
        diagnostics=diagnostics, applied_filters=applied, query_interpretation=interpretation,
        warnings=[],
    )
